torch/distributed/_composable/fsdp/_fsdp_param.py

```diff
@@ -236,6 +254,24 @@ def init_dtype_attrs(self, mp_policy: MixedPrecisionPolicy):
         self.reduce_dtype = reduce_dtype
         # None indicates that the mixed precision is not enabled
 
+    def _init_extensions(self) -> None:
+        inner_tensor = self._sharded_local_tensor
+        has_fsdp_pre_all_gather = hasattr(inner_tensor, "fsdp_pre_all_gather")
+        has_fsdp_post_all_gather = hasattr(inner_tensor, "fsdp_post_all_gather")
+        if has_fsdp_pre_all_gather != has_fsdp_post_all_gather:
+            raise AssertionError(
+                "Both fsdp_pre_all_gather and fsdp_post_all_gather should be defined "
+                f"if using all-gather extensions: {inner_tensor}"
+            )
+        if has_fsdp_pre_all_gather:
+            if self.padded_sharded_param_size != self._sharded_local_tensor.size():
```

Comment:
We make this assumption for now until `DTensor` always pads its local tensor. In particular, we want to run `fsdp_pre_all_gather()` from the ND sharded parameter.